src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.h

```diff
@@ -0,0 +1,270 @@
+  };
+
+  delta_buffer_t delta_buffer;
+  delta_buffer_t *maybe_get_delta_buffer() {
+    return is_mutation_pending() ? &delta_buffer : nullptr;
+  }
+
+  find_lextent_ret find_lextent(ext_context_t ec, objaddr_t lo, extent_len_t len) final;
+
+  insert_ret insert(ext_context_t ec, objaddr_t lo, lext_map_val_t val) final;
+
+  rm_lextent_ret rm_lextent(ext_context_t ec, objaddr_t lo, lext_map_val_t val) final;
+
+  split_children_ret make_split_children(ext_context_t ec) final;
+
+  full_merge_ret make_full_merge(ext_context_t ec, ExtMapNodeRef right) final;
+
+  make_balanced_ret make_balanced(ext_context_t ec, ExtMapNodeRef _right, bool prefer_left) final;
+
+  std::ostream &print_detail_l(std::ostream &out) const final;
+
+  extent_types_t get_type() const final {
+    return type;
+  }
+
+  ceph::bufferlist get_delta() final {
+    assert(!delta_buffer.empty());
+    ceph::buffer::ptr bptr(delta_buffer.get_bytes());
+    delta_buffer.copy_out(bptr.c_str(), bptr.length());
+    ceph::bufferlist bl;
+    bl.push_back(bptr);
+    return bl;
+  }
+
+  void apply_delta(const ceph::bufferlist &_bl) final {
+    assert(_bl.length());
+    ceph::bufferlist bl = _bl;
+    bl.rebuild();
+    delta_buffer_t buffer;
+    buffer.copy_in(bl.front().c_str(), bl.front().length());
+    buffer.replay(*this);
+  }
+
+  bool at_max_capacity() const final {
+    return get_size() == get_capacity();
+  }
+
+  bool at_min_capacity() const {
+    return get_size() == get_capacity() / 2;
+  }
+
+  std::pair<internal_iterator_t, internal_iterator_t> bound(
+    objaddr_t l, objaddr_t r) {
+    auto retl = begin();
+    for (; retl != end(); ++retl) {
+      if (retl->get_next_key_or_max() > l)
+        break;
+    }
+    auto retr = retl;
+    for (; retr != end(); ++retr) {
+      if (retr->get_key() >= r)
+        break;
+    }
+    return std::make_pair(retl, retr);
+  }
+
+  using split_entry_ertr = TransactionManager::read_extent_ertr;
+  using split_entry_ret = split_entry_ertr::future<ExtMapNodeRef>;
+  split_entry_ret split_entry(ext_context_t ec, objaddr_t lo,
+                              internal_iterator_t, ExtMapNodeRef entry);
+
+  using merge_entry_ertr = TransactionManager::read_extent_ertr;
+  using merge_entry_ret = merge_entry_ertr::future<ExtMapNodeRef>;
+  merge_entry_ret merge_entry(ext_context_t ec, objaddr_t lo,
+                              internal_iterator_t iter, ExtMapNodeRef entry);
+
+  internal_iterator_t get_containing_child(objaddr_t lo);
+
+};
+
+/**
+ * ExtMapLeafNode
+ *
+ * Abstracts operations on and layout of leaf nodes for the
+ * ExtentMap Tree.
+ *
+ * Layout (4k):
+ *   num_entries: uint32_t               4b
+ *   meta       : depth                  4b
+ *   (padding)  :                        8b
+ *   keys       : objaddr_t[204]         (204*4)b
+ *   values     : lext_map_val_t[204]    (204*16)b
+ *                                       = 4096
+ */
+constexpr size_t LEAF_NODE_CAPACITY = 204;
```

Comment:
currently fixed_kv_node_layout template can't accept expression, only constant,  here is keep same as lba, and will fix it later.

Reply:
since C++11, `constexpr` is used in lots of places where a constant should be used. you can even use a constexpr function for deducing the array size. see https://en.cppreference.com/w/cpp/language/constexpr

Reply:
fixed. pleased see update.